quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -17,74 +24,109 @@ pub(super) fn run(expression: &Expression) -> Expression {
 /// practical value in anecdotal testing.
-const LIMIT: u64 = 10;
-
-/// Recursively simplify an [`Expression`] by hand, breaking into cases to make things more
-/// manageable.
-fn simplify(e: &Expression, limit: u64) -> Expression {
-    if limit == 0 {
-        // bail
-        e.clone()
-    } else {
-        match e {
+const LIMIT: Limit = Limit(10);
+
+/// We use a separate type for the limit so that we can enforce the use of saturating subtraction,
+/// but we don't use [`std::num::Saturating`] so that we can write that in terms of literals.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+struct Limit(u64);
+
+impl std::cmp::PartialEq<u64> for Limit {
+    #[inline(always)]
+    fn eq(&self, other: &u64) -> bool {
+        self.0 == *other
+    }
+}
+
+impl std::ops::Sub<u64> for Limit {
+    type Output = Self;
+
+    #[inline(always)]
+    fn sub(self, rhs: u64) -> Self::Output {
+        Limit(self.0.saturating_sub(rhs))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Simplifier {
+    memo_table: HashMap<ArcIntern<Expression>, ArcIntern<Expression>>,
+}
+
+impl Simplifier {
+    fn new() -> Self {
+        Self {
+            memo_table: HashMap::new(),
+        }
+    }
+
+    #[allow(dead_code)]
+    fn debug_memo_table(&self, w: &mut impl std::io::Write) -> std::io::Result<()> {
+        use crate::quil::Quil as _;
+
+        writeln!(w, "Simplifier memo table:")?;
+        for (k, v) in &self.memo_table {
+            writeln!(
+                w,
+                "    {} => {}",
+                k.to_quil_or_debug(),
+                v.to_quil_or_debug(),
+            )?
+        }
+
+        Ok(())
+    }
+
+    /// Recursively simplify an [`Expression`] by hand, breaking into cases to make things more
+    /// manageable.  Terminates the simplification when `limit` reaches `0`.
+    ///
+    /// The `limit` decreases whenever we take a step deeper within the expression.  The special
+    /// case simplification functions, the ones that do all the work, don't need to decrement the
+    /// limit at the top of the function.
+    ///
+    /// Invariant: Never returns [`Expression::PiConstant`].
+    fn simplify(&mut self, e: ArcIntern<Expression>, limit: Limit) -> ArcIntern<Expression> {
+        if let Some(simplified) = self.memo_table.get(&e) {
+            return simplified.clone();
+        }
+
+        let result = match e.as_ref() {
+            // Even at a limit of 0, replace `pi` with 3.14….  This gets us the extremely useful
+            // invariant that `Expression::PiConstant` can never be returned from `simplify`.
+            Expression::PiConstant => ArcIntern::new(Expression::Number(PI)),
+
+            // We're out of gas; this is as simplified as things get
+            _ if limit == 0 => e.clone(),
+
+            // Atoms don't need to be simplified
             Expression::Address(_) | Expression::Number(_) | Expression::Variable(_) => e.clone(),
```

Comment:
if these don't need simplification can this go above the limit check?

Reply:
Yeah, it's six of one, I can move them

Reply:
oh, oops, I see what you mean